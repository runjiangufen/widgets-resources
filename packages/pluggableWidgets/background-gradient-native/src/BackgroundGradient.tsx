import { ReactElement, createElement } from "react";
import { Pressable, NativeModules, Alert } from "react-native";
import LinearGradient from "react-native-linear-gradient";
import { all } from "deepmerge";
import { executeAction } from "@mendix/piw-utils-internal";
import defaultStyle, { CustomStyle } from "./ui/Styles";
import { BackgroundGradientProps } from "../typings/BackgroundGradientProps";

export type props = BackgroundGradientProps<CustomStyle>;

const opacityValidation = (opacity: number | undefined): number => {
    if (opacity === undefined) {
        return defaultStyle.opacity / 100;
    }
    const opacityVal = Number(opacity);
    if (isNaN(opacityVal)) {
        throw new Error("Opacity must be a number.");
    }
    if (opacityVal < 0 || opacityVal > 100) {
        console.warn("Opacity must be between 0 and 100.");
    }

    return opacityVal / 100;
};

const angleValidation = (angle: number | undefined): number => {
    if (angle === undefined) {
        return defaultStyle.angle;
    }
    const angleVal = Number(angle);
    if (isNaN(angleVal)) {
        throw new Error("Angle must be a number.");
    }

    return angleVal;
};

export function BackgroundGradient({ name, colorList, content, onClick, style }: props): ReactElement {
    if (!("BVLinearGradient" in NativeModules.UIManager)) {
        Alert.alert("", "The widget 'Background gradient' requires an updated 'Make It Native 9' application");
    }

    const styles = all<CustomStyle>([defaultStyle, ...style]);
    const angle = angleValidation(styles.angle);
    const opacity = opacityValidation(styles.opacity);

    let sortedColorList = (styles.colorList && colorList.length === 0 ? styles.colorList : colorList).sort(
        (a, b) => Number(a.offset) - Number(b.offset)
    );

    if (sortedColorList.length === 0) {
        throw new Error("The color list could not be empty.");
    }

    // checking if the color list only have one color item , then we should duplicate it. one color list throws an exception on android.
    if (sortedColorList.length === 1) {
        sortedColorList = [...sortedColorList, ...sortedColorList];
    }

    const colors = sortedColorList.map(colorsObject => colorsObject.color.toLowerCase());
    const offsets = sortedColorList.map(colorsObject => Number(colorsObject.offset));

    return (
        <Pressable
            onPress={() => {
                executeAction(onClick);
            }}
            testID={name}
            style={({ pressed }) => ({
                flex: styles.container.flex,
                opacity: onClick?.canExecute && pressed ? opacity * 0.3 : opacity
            })}
        >
            <LinearGradient colors={colors} locations={offsets} useAngle angle={angle} style={styles.container}>
                {content}
            </LinearGradient>
        </Pressable>
    );
}
