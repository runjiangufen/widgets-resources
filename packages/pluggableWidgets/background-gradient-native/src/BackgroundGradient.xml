<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.backgroundgradient.BackgroundGradient" pluginWidget="true" needsEntityContext="true" offlineCapable="true" supportedPlatform="Native" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Background gradient</name>
    <description />
    <studioProCategory>Structure</studioProCategory>
    <studioCategory>Structure</studioCategory>
    <icon />
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="colorList" type="object" isList="true" required="false">
                    <caption>Color list</caption>
                    <description />
                    <properties>
                        <propertyGroup caption="Item">
                            <property key="color" type="string" required="false">
                                <caption>Color</caption>
                                <description>Valid color value format: #d0d0d0, rgb(115,155,155), or rgba(195,226,226,0.5)</description>
                            </property>
                            <property key="offset" type="decimal" defaultValue="0.0">
                                <caption>Offset</caption>
                                <description>A color-stop's value, followed by one or more optional stop positions. Should be between 0.0 and 1.0</description>
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="content" type="widgets">
                    <caption>Content</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Visibility">
                <systemProperty key="Visibility" />
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Events">
            <property key="onClick" type="action" required="false">
                <caption>On click</caption>
                <description />
            </property>
        </propertyGroup>
    </properties>
</widget>
