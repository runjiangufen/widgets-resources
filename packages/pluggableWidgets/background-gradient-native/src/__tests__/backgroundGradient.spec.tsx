import { createElement } from "react";
import { Text } from "react-native";
import { render, fireEvent } from "@testing-library/react-native";
import Big from "big.js";
import { ActionValue } from "mendix";
import { BackgroundGradient, props } from "../BackgroundGradient";

describe("Background gradient", () => {
    let defaultProps: props;
    beforeEach(() => {
        defaultProps = {
            colorList: [
                { color: "#fff", offset: Big(0) },
                { color: "#000", offset: Big(1) }
            ],
            content: <Text>Test</Text>,
            name: "test",
            style: []
        };
    });
    it("render correctly", () => {
        const component = render(<BackgroundGradient {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("render with undefined opacity", () => {
        defaultProps = {
            ...defaultProps,
            style: [{ container: {}, opacity: undefined }]
        };
        const component = render(<BackgroundGradient {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("render with undefined angle", () => {
        defaultProps = {
            ...defaultProps,
            style: [{ container: {}, angle: undefined }]
        };
        const component = render(<BackgroundGradient {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("should throw error when colorList is empty", () => {
        defaultProps = {
            ...defaultProps,
            colorList: []
        };
        expect(() => render(<BackgroundGradient {...defaultProps} />)).toThrowError();
    });
    it("should trow error when opacity is not a number", () => {
        defaultProps = {
            ...defaultProps,
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            style: [{ container: {}, opacity: "test" }]
        };
        expect(() => render(<BackgroundGradient {...defaultProps} />)).toThrowError();
    });
    it("should trow error when angle is not a number", () => {
        defaultProps = {
            ...defaultProps,
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            style: [{ container: {}, angle: "test" }]
        };
        expect(() => render(<BackgroundGradient {...defaultProps} />)).toThrowError();
    });
    it("is should console warn when opacity is not between 0 and 100", () => {
        defaultProps = {
            ...defaultProps,
            style: [{ container: {}, opacity: 101 }]
        };
        const spy = jest.spyOn(console, "warn");
        render(<BackgroundGradient {...defaultProps} />);
        expect(spy).toHaveBeenCalled();
    });
    it("render background gradient with custom style", () => {
        defaultProps = {
            ...defaultProps,
            colorList: [],
            style: [
                {
                    angle: 0,
                    opacity: 100,
                    container: {},
                    colorList: [
                        { color: "#fff", offset: Big(0) },
                        { color: "#000", offset: Big(1) }
                    ]
                }
            ]
        };
        const component = render(<BackgroundGradient {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("render background gradient with one color", () => {
        defaultProps = {
            ...defaultProps,
            colorList: [{ color: "#fff", offset: Big(0) }]
        };
        const component = render(<BackgroundGradient {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("call onClick when the container is pressed", () => {
        const actionExecution = jest.fn();
        const action: ActionValue = {
            canExecute: true,
            execute: actionExecution,
            isExecuting: false
        };
        const component = render(<BackgroundGradient {...defaultProps} onClick={action} />);
        const pressable = component.getByTestId("test");
        fireEvent.press(pressable);
        expect(actionExecution).toBeCalled();
    });
});
