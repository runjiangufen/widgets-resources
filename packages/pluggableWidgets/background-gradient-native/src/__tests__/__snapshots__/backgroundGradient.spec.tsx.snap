// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background gradient render background gradient with custom style 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": undefined,
      "opacity": 1,
    }
  }
  testID="test"
>
  <BVLinearGradient
    angle={0}
    colors={
      Array [
        4294967295,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        NaN,
        NaN,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={Object {}}
    useAngle={true}
  >
    <Text>
      Test
    </Text>
  </BVLinearGradient>
</View>
`;

exports[`Background gradient render background gradient with one color 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": undefined,
      "opacity": 1,
    }
  }
  testID="test"
>
  <BVLinearGradient
    angle={0}
    colors={
      Array [
        4294967295,
        4294967295,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        0,
        0,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={Object {}}
    useAngle={true}
  >
    <Text>
      Test
    </Text>
  </BVLinearGradient>
</View>
`;

exports[`Background gradient render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": undefined,
      "opacity": 1,
    }
  }
  testID="test"
>
  <BVLinearGradient
    angle={0}
    colors={
      Array [
        4294967295,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        0,
        1,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={Object {}}
    useAngle={true}
  >
    <Text>
      Test
    </Text>
  </BVLinearGradient>
</View>
`;

exports[`Background gradient render with undefined angle 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": undefined,
      "opacity": 1,
    }
  }
  testID="test"
>
  <BVLinearGradient
    angle={0}
    colors={
      Array [
        4294967295,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        0,
        1,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={Object {}}
    useAngle={true}
  >
    <Text>
      Test
    </Text>
  </BVLinearGradient>
</View>
`;

exports[`Background gradient render with undefined opacity 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": undefined,
      "opacity": 1,
    }
  }
  testID="test"
>
  <BVLinearGradient
    angle={0}
    colors={
      Array [
        4294967295,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        0,
        1,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={Object {}}
    useAngle={true}
  >
    <Text>
      Test
    </Text>
  </BVLinearGradient>
</View>
`;
